File: storage/data_transpiled/at_coder/arc015/D/2201657.java
public static double getDoubleDouble ( ) {
  int T = input . nextInt ( ) ;
  int N = input . nextInt ( ) ;
  double P = Double . parseDouble ( input . nextLine ( ) ) ;
  double [ ] [ ] D = new double [ N ] [ N ] ;
  for ( int i = 0 ;
  i < N ;
  i ++ ) {
    double q , x , t ;
    q = Double . parseDouble ( input . nextLine ( ) ) ;
    x = Double . parseDouble ( input . nextLine ( ) ) ;
    t = Double . parseDouble ( input . nextLine ( ) ) ;
    D [ i ] = new double [ ] {
      t , x , q }
      ;
    }
    Arrays . sort ( D ) ;
    double cur = 1 - P + P * Math . pow ( x * q , 2 ) ;
    int j = 0 ;
    double ans = 0. ;
    double res = 1. ;
    for ( int y = 0 ;
    y < T ;
    y ++ ) {
      while ( j < N && D [ j ] [ 0 ] <= y ) {
        double t , x , q ;
        cur += P * q * ( 1 - x ) ;
        j ++ ;
      }
      ans += res ;
      res *= cur ;
    }
    System . out . println ( "" + ans + "f" ) ;
    return ans ;
  }
  

File: storage/data_original/at_coder/arc015/D/2201657.py
# seishin.py
T, N, P = input().split()
T = int(T); N = int(N); P = float(P)
D = []
for i in range(N):
    q, x, t = input().split()
    q = float(q); x = int(x); t = int(t)
    D.append((t, x, q))
D.sort()

# ??????
# M[p] := ??p?????? (????????????p???????)
# M[0] = 1.0
# M[1] = (1 - P) + P*(x_0*q_0 + ..., + x_{N-1}*q_{N-1})
#   (=> cur????)
# ...
# (??t_i????????? t_i < x ?????)
# M[y] = <???????????> + <(????????)??i???> + <(??????????)??i???>
#      = (1 - P) + P*(q_0*1 + ... + q_i*1) + P*(x_{i+1}*q_{i+1} + ... + x_{N-1}*q_{N-1})
#
# ??i?????????? ?_{y=0}^{i} M[y] ???
#
# ????
#  ans = ?_{t=0}^{T-1} ?_{y=0}^{t-1} M[y]
# ??????????
cur = 1 - P + P*sum(x*q for t, x, q in D)
j = 0
ans = 0.
res = 1.
for y in range(T):
    while j < N and D[j][0] <= y:
        # ??j????????? P*x_j*q_j ???? P*q_j ???
        t, x, q = D[j]
        cur += P*q*(1-x)
        j += 1
    ans += res # ans += ?_{y=0}^{i} M[y]
    res *= cur # res *= M[y]
# ?????10^-3????OK
print("%.08f" % ans)